runners/direct-java/src/main/java/org/apache/beam/runners/direct/SplittableProcessElementsEvaluatorFactory.java

```diff
@@ -0,0 +1,143 @@
+import org.apache.beam.sdk.util.state.StateInternalsFactory;
+import org.apache.beam.sdk.util.state.TimerInternalsFactory;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.PCollectionTuple;
+import org.apache.beam.sdk.values.TupleTag;
+import org.joda.time.Instant;
+
+class SplittableProcessElementsEvaluatorFactory<InputT, OutputT, RestrictionT>
+    implements TransformEvaluatorFactory {
+  private final ParDoEvaluatorFactory<
+          KeyedWorkItem<String, ElementAndRestriction<InputT, RestrictionT>>, OutputT>
+      delegateFactory;
+  private final EvaluationContext evaluationContext;
+
+  SplittableProcessElementsEvaluatorFactory(EvaluationContext evaluationContext) {
+    this.evaluationContext = evaluationContext;
+    this.delegateFactory = new ParDoEvaluatorFactory<>(evaluationContext);
+  }
+
+  @Override
+  public <T> TransformEvaluator<T> forApplication(
+      AppliedPTransform<?, ?, ?> application, CommittedBundle<?> inputBundle) throws Exception {
+    @SuppressWarnings({"unchecked", "rawtypes"})
+    TransformEvaluator<T> evaluator =
+        (TransformEvaluator<T>)
+            createEvaluator((AppliedPTransform) application, (CommittedBundle) inputBundle);
+    return evaluator;
+  }
+
+  @Override
+  public void cleanup() throws Exception {
+    delegateFactory.cleanup();
+  }
+
+  @SuppressWarnings({"unchecked", "rawtypes"})
+  private TransformEvaluator<KeyedWorkItem<String, ElementAndRestriction<InputT, RestrictionT>>>
+      createEvaluator(
+          AppliedPTransform<
+                  PCollection<KeyedWorkItem<String, ElementAndRestriction<InputT, RestrictionT>>>,
+                  PCollectionTuple, SplittableParDo.ProcessElements<InputT, OutputT, RestrictionT>>
+              application,
+          CommittedBundle<InputT> inputBundle)
+          throws Exception {
+    final SplittableParDo.ProcessElements<InputT, OutputT, RestrictionT> transform =
+        application.getTransform();
+
+    DoFnLifecycleManager fnManager = delegateFactory.getFnClone(transform.getProcessFn());
+
+    SplittableParDo.ProcessFn<InputT, OutputT, RestrictionT, ?> processFn =
+        ((SplittableParDo.ProcessFn<InputT, OutputT, RestrictionT, ?>) fnManager.get());
+
+    String stepName = evaluationContext.getStepName(application);
+    final DirectExecutionContext.DirectStepContext stepContext =
+        evaluationContext
+            .getExecutionContext(application, inputBundle.getKey())
+            .getOrCreateStepContext(stepName, stepName);
+
+    ParDoEvaluator<KeyedWorkItem<String, ElementAndRestriction<InputT, RestrictionT>>, OutputT>
+        parDoEvaluator =
+            delegateFactory.createParDoEvaluator(
+                application,
+                transform.getSideInputs(),
+                transform.getMainOutputTag(),
+                transform.getSideOutputTags().getAll(),
+                stepContext,
+                fnManager);
+
+    processFn.setStateInternalsFactory(
```

Comment:
The `ProcessFn` is to all appearances a System DoFn, as it requires intervention from the system to be properly executed. As such, you should construct it on demand, as it's an implementation detail of `SplittableDoFn` - which means you can construct a new instance within the evaluator for each bundle. The `SplittableParDo` should instead carry as its payload the `SplittableDoFn`, and the `ProcessFn` can be constructed with appropriate `StateInternals`, `TimerInternals`, and `OutputWindowedValue` (`WindowedOutputReciever`?). This is almost the same pattern we formerly used for `GroupAlsoByWindowsDoFn` before that code was entirely migrated to be entirely a primitive in https://github.com/apache/incubator-beam/pull/1251

You can still invoke the methods via `ParDoEvaluator` (though I think generally that's probably discouraged, as the runner will probably want to interact with a `SplittableDoFn` in ways that can't easily be expressed in terms of a `ParDo`).